Add commit message guideline document (#138)
canonical-webteam/practices: workflow/commit-messages.md

> # Commit Guidelines

Reviewer: Since we Jekyllised this repo, now documents should include the title and a description in YAML frontmatter:

``` markdown
---
title: Commit guidelines
description: {Come up with a nice description}
---

... the rest of the document
```

Also, we use sentence case rather than title case for titles, so "guidelines" should have a small "g". (Somewhat obscurely mentioned in [this practice](https://github.com/canonical-webteam/practices/blob/master/content/copy-reviews.md#checklist)).

> As a general guideline you can follow the idea that `a single commit should enclose a single problem/feature that has been fixed/added`.

Reviewer: Backticks will place `<code>` around the block. Please don't use them for emphasis or quotes. Either use `"` double-quotes or `*` emphasis.

This way of organizing your changes to the codebase makes is easy for reviewers and people who try to understand the history of a project to determine what has been happening in the project.

Here a both a good and a bad example of of commit message.

> ### BAD

Reviewer: Here, you're using `<h3>`s under an `<h1>` above. This is usually a sign that the documents structure could be reconsidered. In this case, I think maybe "Good" and "Bad" shouldn't be headings (and probably shouldn't be in all capitals). In this case, I think just:

``` markdown
Good:

{example}

Bad:

{example}
```

would have been fine. Or maybe even just write it as a sentence:

> For example, "Fix bug where database gets deleted on POST request" is better than "fixes bug".

```
fixes bug
```
### GOOD
```
Fix bug where database gets deleted on POST request
```

## Breaking up large tickets

If your feature/ticket includes mutliple things that you have to complete in order to submit a pull request, these should all be seperate commit messages, according to the general guideline above. This will make it easy for the reviewer to understand what exactly has happened.

For adding a new page in an example Project this could look like the following
```
Add the Person model
Configure the Views to use the new Person Model
Use the Data from the Views in the index template
```

> Further information about how to write good commit messages can be found [here](https://chris.beams.io/posts/git-commit/)

Reviewer: When linking, it would be nice to include more descriptive link text, rather than "here" or "click here". Maybe this sentence could be restructured to accommodate that as something like:

> Chris Beams has written [a good article on how to write commit messages](https://chris.beams.io/posts/git-commit/).

This also achieves giving credit to the author, which is supposed to be one of our values. (And maybe this linking point should be added to a new document about writing for the web at some point).

## Rebasing

> If it is possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer 2: I would just say `If possible` instead of `If it is possible`.

---
title: Commit guidelines
description: Some general guidelines to help you in writing good commits for collaboration and clarity.
---

> # Commit Guidelines

Reviewer: Actually, if you have `title: Commit guidelines`, I think you should remove this heading, 'cos I think it will look duplicated.

You can see how it looks with `gem install jekyll jekyll-redirect-from; jekyll serve`, and also see [other documents](https://github.com/canonical-webteam/practices/blob/master/project-structure/server-side-frameworks.md) for examples.

Author: Good catch. Wrote this outside the repo first.

As a general guideline you can follow the idea that _a single commit should enclose a single problem/feature that has been fixed/added_.

This way of organizing your changes to the codebase makes is easy for reviewers and people who try to understand the history of a project to determine what has been happening in the project.

Here a both a good and a bad example of of commit message.

**Bad:**

```
fixes bug
```

**Good:**

```
Fix bug where database gets deleted on POST request
```

## Breaking up large tickets

> If your feature/ticket includes mutliple things that you have to complete in order to submit a pull request, these should all be seperate commit messages, according to the general guideline above. This will make it easy for the reviewer to understand what exactly has happened.

Reviewer 3: > these should all be seperate commit messages

Do you actually mean seperate commit messages, or multi-line commit messages?

Reviewer 4: As far as I understand the intention was for it to be separate commits (with proper messages).

I agree that it's usually tricky. We don't often have a perfect plan and errorless solution in mind that nicely maps in to a clean, logically split commit history. We implement a part of solution, commit it, maybe discover an error later, get some review comment. Keeping commit history clean with always valuable "feature" commits would require constant rebases with a sole purpose of having a clean history - history we will rarely look at.

If the PR itself is small enough (fixes one bug, adds one feature) it can theoretically work in history as one commit. Even if developed in couple of steps it's the final result that counts. When reviewing I rarely look at individual commits. It's the whole PR diff that I review.

While I agree that having meaningful commits with good commit messages is valuable, I wouldn't go out of the way cleaning up history, rebasing review comment fixes into past individual commits, etc.

Reviewer: I am a bit concerned with this document that it will make people feel like they *have to* spend a lot of effort tidying commits, and I do agree that this is not worth spending a lot of time on.

Having said that, I do think this is a *little* blaze:

> Keeping commit history clean with always valuable "feature" commits would require constant rebases with a sole purpose of having a clean history - history we will rarely look at.

Clean history is definitely useful. While it's true that the vast majority of history probably won't be inspected, it's also true that people very often do inspect history. I definitely do. And I also link to commits. I tell the community in bug reports e.g. "this feature was added in this commit", or I quickly skim the commits between two dates to see what happened. Or sometimes I cherry-pick commits to quickly grab functionality from other projects. Creating neat, atomic commits that describe individual features that are added are definitely not a waste of time.

So the middle ground I think it would be good to achieve with this document is to outline what would be the ideal way to create commits, but making it clear that if you find it difficult it is not a hard requirement. I *never* want PRs to be blocked because of the format of commit messages - @b-m-f please could you add something to the document to make this clear?

It would certainly be helpful if all our devs would gradually learn how to quickly and efficiently manage Git commits such that it's second nature and doesn't take them much time, but this upskilling will always be a gradual process.

## Tricks for keeping commits tidy

Having said all that, personally, I do my best to rebase commits quite regularly. It's become so second nature that I don't think it slows me down in any significant way. A few simple tricks help.

1. (I think everyone does this), frequently *rebase* (don't merge) your work on top of the latest version of the branch you're merging into. Do this at every opportunity, and ideally just before the PR is merged.
2. Heavy use of `git commit --amend --no-edit`. If you do new work that could simply fit into the previous commit, just do `git add {changed filed}` followed by `git commit --amend --no-edit`
3. Heavy use of interactive rebasing (@tbille showed me this trick, I didn't know it before). Since this is fairly involved, I'll explain it below:

### Interactive rebasing

Let's say:

1. You do some work in `users.js` and you commit it with "Improve user logic"
2. You do some work in `README.md` and you commit it with "Explain user logic better in README"
3. Now you realise you need another fix to your `users.js` work. I want it to be added to my "Improve user logic" commit, so I do this:

``` bash
$ git add users.js
$ git commit -m '... to be rebased'
$ git rebase -i HEAD^^^  # Interactively rebase the last 3 commits
```

This will open an editor with this contents:

```
pick 239d4c3 Improve user logic
pick adc7c21 Explain user logic better in README
pick fd4f81b ... to be rebased
```

Now if I reorder my commits to move line 3 to line 2, and replace "pick" with "squash" (or simply "s"), then it will achieve what I want:

```
pick 239d4c3 Improve user logic
squash fd4f81b ... to be rebased
pick adc7c21 Explain user logic better in README
```

Save and exit, and then it will open another editor to ask you what you want the new commit message for the merged commit to be:

```
# This is a combination of 2 commits.
# This is the 1st commit message:

Improve user logic

# This is the commit message #2:

... to be rebased
```

But in this case I probably just want to delete the second message. Save and exist, and voila:

``` bash
$ git log
commit b87706ed594c983841857b51923e499988760d41 (HEAD -> master)
Author: Robin Winslow <robin@robinwinslow.co.uk>
Date:   Fri Dec 7 11:05:54 2018 +0000

    Explain user logic better in README

commit da1e9d47073cdc3bf86d3658b2e020c7e37292c0
Author: Robin Winslow <robin@robinwinslow.co.uk>
Date:   Fri Dec 7 11:05:40 2018 +0000

    Improve user logic
```

This process might seem complicated, but once you've done it a couple of times, you'll see that it actually doesn't take long. It probably takes me usually about 15-20 seconds.

There are of course some cases where commits can't be neatly re-ordered (although fewer than you'd think). In these cases, after you perform your interactive rebase, you'll get a conflict which you'll have to resolve in the normal way. You'll quickly get a feel for which commits can be reordered and which can't, and if you think it's going to be too much trouble, just don't bother starting. Or `git rebase --abort`.

Reviewer 4: `git rebase -i` is great once you get used to it. I do use it quite regularly on a larger PRs to turn the long list of commits into something more managable (mostly by squash and fixup - as reordering is tricky as @nottrobin pointed out).

@nottrobin That is a great write down here. It would be good to turn it into a document itself or a part of one about git/commit/github process.

Reviewer: Thanks @bartaz. I realise that at this point I've basically written a blog post. I should probably publish it =)

Reviewer 5: @nottrobin You should!

Reviewer: Thank @b-m-f. I've tidied the Wiki post a bit and renamed it:

https://github.com/canonical-webteam/practices/wiki/Tips-on-managing-commit-history-in-Git

I think it would be good to link to it from this document.

Reviewer: I'm turning this into a blog post - the draft is here: https://docs.google.com/document/d/1gPBV9NBPCwVDcEo6mRZSF--WmDgoBa5hL3Uzzc0aZTI/edit

Please feel free to jump in and make any suggestions.

Reviewer 3: @b-m-f I think the language here is still ambiguous.

Reviewer 3: >Clean history is definitely useful. While it's true that the vast majority of history probably won't be inspected, it's also true that people very often do inspect history. I definitely do. And I also link to commits. I tell the community in bug reports e.g. "this feature was added in this commit", or I quickly skim the commits between two dates to see what happened. Or sometimes I cherry-pick commits to quickly grab functionality from other projects. Creating neat, atomic commits that describe individual features that are added is definitely not a waste of time.

Absolutely, but I think we're conflating clean history on master with clean history in a PR, which is what @bartaz was talking about.

Reviewer: I published the blog post: https://blog.ubuntu.com/2018/12/12/tricks-for-keeping-a-tidy-git-commit-history

Reviewer: > Absolutely, but I think we're conflating clean history on master with clean history in a PR, which is what @bartaz was talking about.

For all our repositories on GitHub (except maybe build.snapcraft.io, which we should maybe look into), history in a PR becomes history in master, because we "merge" rather than "squash" (In GitHub, even if you "squash", by default each individual commit message is still included in the overall commit message of the squash commit).

@b-m-f it may be worth just making it clear that caring about the commit history in your PR is largely irrelevant if the repository you're working on squashes commits when PRs are merged. With a note that, in this scenario, you should try to be as detailed as possible in your squash commit.

Author: @squidsoup I have added another sentence for the squashing. Do you think this makes it  less ambiguous?

For adding a new page in an example Project this could look like the following

```
Add the Person model
Configure the Views to use the new Person Model
Use the Data from the Views in the index template
```

For more information about how to write good commit messages you can check out [this great article by Chris Beams](https://chris.beams.io/posts/git-commit/)

## Rebasing

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer: Although I'm not sure if there should be "a team standard" (there may be project standards), but personally I much prefer the GitHub default of "Merge commit", which uses a standard "git merge" to merge a feature branch with master.

Using "squash" loses a significant amount of information about how the work had been split up. We are supposed to try to keep PRs as small as possible, but there are definitely mixed results with this, and even if you do this well there are cases where adding one whole piece of functionality requires changes in quite a number of places.

If someone has taken the time to separate out the creation of the "models.User" from "views.login" as 2 separate commits (as I'd hope they might), I don't want to lose that separation by squashing them together.

This does somewhat depend on how good we are at keeping our commits neat. Obviously I'd rather have 1 squashed commit for the whole PR than 30 commits called "fixed typo". But for my PRs, I'd certainly rather the history was preserved as I wrote it.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer 4: If we are talking about GitHub 'Squash on merge' feature it simply squashes whole PR into one commit and rebases it on top of master (or whatever branch you are merging to). As far as I remember (we use it by default in [build.snapcraft.io](https://github.com/canonical-websites/build.snapcraft.io)) it creates a commit message that lists all the messages of commits being squashed.

I believe we had a dev catch up discussion about merging vs rebasing vs squashing PRs and the consensus was to use merge by default not to loose commit history by squashing whole PR. Personally I don't think we loose much by squashing, although it may end up with huge commit diffs in large PRs.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer: > If we are talking about GitHub 'Squash on merge' feature it simply squashes whole PR into one commit and rebases it on top of master (or whatever branch you are merging to). As far as I remember (we use it by default in build.snapcraft.io) it creates a commit message that lists all the messages of commits being squashed.

Yes that's right. The lost information I was referring to is the information about which commit message pertains to which part of the diff, which to me seems pretty important (imagine if we disassociated all of git's commit messages from the diffs), although it does depend on whether the person filing the PR cared about keeping an accurate history.

> I believe we had a dev catch up discussion about merging vs rebasing vs squashing PRs and the consensus was to use merge by default not to loose commit history by squashing whole PR.

Yes I remember that discussion. I agree my memory is that the preference of the team to use merge, which is also my preference. I'm not sure we made a very strong insistence though. We could consider this again, or we could continue with that standard.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Author: @squidsoup @bartaz @nottrobin , Should I leave this sentence in here?  The squashing vs merging probably fits better into the github workflow/Pull Request document?

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer: @squidsoup since this sentence isn't actually related to merging PRs, AFAICT it's only about keeping the commits in your feature branch  tidy, are you happy with its current form? Or would you like to suggest an alternative?

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer 3: > @squidsoup Or would you like to suggest an alternative?

I personally don't feel there's much to be gained by investing energy in making PRs with commits per feature in large branches, particularly when our workflow for two of our large projects, CRBS and MAAS squash commits on merge. History on master should be clean, but I would rather spend time on the code rather than crafting feature commits when those get lost anyway.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer 3: > > If we are talking about GitHub 'Squash on merge' feature it simply squashes whole PR into one commit and rebases it on top of master (or whatever branch you are merging to). As far as I remember (we use it by default in build.snapcraft.io) it creates a commit message that lists all the messages of commits being squashed.
> 
Github may do that, but LP does not. Manually writing a multiline commit message for large branches isn't difficult however.

> Yes that's right. The lost information I was referring to is the information about which commit message pertains to which part of the diff, 

I would suggest that if you have to break up your code review by commit, the author should be making three PRs and the branch is too large.

> > I believe we had a dev catch up discussion about merging vs rebasing vs squashing PRs and the consensus was to use merge by default not to loose commit history by squashing whole PR.

Why is intermediary history valuable? Personally speaking, most of my intermediary commits are junk, and little more than placemarkers.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer: > @squidsoup: I personally don't feel there's much to be gained by investing energy in making PRs with commits per feature in large branches, particularly when our workflow for two of our large projects, CRBS and MAAS squash commits on merge. History on master should be clean, but I would rather spend time on the code rather than crafting feature commits when those get lost anyway.

> Why is intermediary history valuable, bearing in mind our workflow on LP? Personally speaking, most of my intermediary commits are junk, and little more than placemarkers.

I completely agree, you should make a call about what makes sense for working on CRBS and MAAS. For almost all of us, this isn't "intermediary history", it's simply "history". But I can see that for you it's different, and it is definitely a waste of time to worry about history that's going to be thrown away.

Firstly, in my opinion this document should only be considered advisory (as I've mentioned a few times) and so you or anyone else should be absolutely free to ignore any of its advice that doesn't work for you.

I assume MAAS and CRBS work this way because we don't control those codebases? So we can assume that we'd try to follow these guidelines for codebases we *do* control? @b-m-f it would be worth, I think, explicitly adding a concession for when we're working on codebases that we don't control.

> I would suggest that if you have to break up your code review by commit, the author should be making three PRs and the branch is too large.

This I completely disagree with. One *could* consider that Git history should merely be a history of all the merged PRs, but this offers *much* less detail than I'm used to working with. PRs should be as small as possible, but at the very least a PR is likely to include several things - e.g. some code, an update to a README, and some tests. I think it's much more useful, if you can, to see PRs as encapsulating whole new pieces of functionality, and commits as describing individual changes.

This inevitably gives you much more detailed history & the commits will have much smaller diffs. If you've ever tried working through history where people have created huge commits I think you'll appreciate how frustrating it can be.

E.g. my PR might have:

```
102: Document user logic
101: Test user logic
100: Improve user logic
```

If I "merge" this PR into master, then my master history will look like:

```
103: Merge PR #56: New user logic (merge of commit 102 & commit 99)
- 102: Document user logic
- 101: Test user logic
- 100: Improve user logic
99: Merge PR #55: Add secret cryptocurrency miner
```

And in GitHub those commits will even tell you that they came from that PR.

If there's no other choice, simply one commit per PR is acceptable, but we can have more detail, I think that's always better.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer 3: > I assume MAAS and CRBS work this way because we don't control those codebases?

Right, and potentially out of our hands, that's just the way landings work for MAAS and CRBS (RBAC) and I think it's probably best if we try to conform to their workflows as much as possible. I think @b-m-f's suggestions are great, but we probably just need a caveat that they only apply to projects that we have complete control over landing and ci.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Author: So, this has been lying around quite a while. 

Do you think that the sentence 
```
But make sure that this is a something that is done throughout your project. If another workflow is used, such as squashing all your commits together before you merge your branch into master this is not necessary. 
```
addresses this concern?

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer 3: @b-m-f sure, that's fine.

> If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

Reviewer: Yep sounds good to me.

As a general guideline you can follow the idea that _a single commit should enclose a single problem/feature that has been fixed/added_.

This way of organizing your changes to the codebase makes is easy for reviewers and people who try to understand the history of a project to determine what has been happening in the project.

Here a both a good and a bad example of of commit message.

**Bad:**

```
fixes bug
```

**Good:**

```
Fix bug where database gets deleted on POST request
```

## Breaking up large tickets

If your feature/ticket includes mutliple things that you have to complete in order to submit a pull request, these should all be seperate commit messages, according to the general guideline above. This will make it easy for the reviewer to understand what exactly has happened.
But make sure that this is a something that is done throughout your project. If another workflow is used, such as squashing all your commits together before you merge your branch into master this is not necessary. Just make sure that your commit message for the squash commit includes a detailed message of the changes.

> 

Reviewer 3: Might be good to mention here that it is fine to have multiline commits (in the sqashed commits workflow) in the form, ensuring all the changes are enumerated:

  _Overall commit message._

  * _Change 1_
  * _Change 2_

Author: What do you think of the content I have added @squidsoup @nottrobin . Should it be be more explicit to use multiline commit messages?

---
title: Commit guidelines
description: Some general guidelines to help you in writing good commits for collaboration and clarity.
---

> This document was written as an overview of how an optimal commit should be written. But in the end it is still your decision , no Pull Request should be stuck in a review phase because its commits differ from the practices described below.

Reviewer 5: ```suggestion
This document was written as an overview of how an optimal commit should be written. But in the end it is still your decision, no Pull Request should be stuck in a review phase because it's commits differ from the practices described below.
```

> But if you are unsure of how to write them, thos should give you a good starting point for great collaboration.

Reviewer 5: ```suggestion
But if you are unsure of how to write good commit messages, this guide should give you a good starting point for great collaboration.
```

> But if you are unsure of how to write them, thos should give you a good starting point for great collaboration.

Reviewer 6: thos > this

Author: Thanks :) Fixed

> As a general guideline you can follow the idea that _a single commit should enclose a single problem/feature that has been fixed/added_.

Reviewer 5: ```suggestion
As a general guide, you can follow the idea that _a single commit should enclose a single problem/feature that has been fixed/added_.
```

This way of organizing your changes to the codebase makes is easy for reviewers and people who try to understand the history of a project to determine what has been happening in the project.

> Here a both a good and a bad example of of commit message.

Reviewer 5: ```suggestion
Here is both a good and bad example of a commit message:
```

**Bad:**

```
fixes bug
```

**Good:**

```

> Fix bug where database gets deleted on POST request

Reviewer 6: add period

Author: Thanks :) Fixed

```

## Breaking up large tickets

> If your feature/ticket includes mutliple things that you have to complete in order to submit a pull request, these should all be seperate commit messages, according to the general guideline above. This will make it easy for the reviewer to understand what exactly has happened.

Reviewer 5: ```suggestion
If your feature/ticket includes multiple things that you have to complete in order to submit a pull request, these should all be separate commit messages, according to the general guideline above. This will make it easy for the reviewer to understand what exactly has happened.
```

But make sure that this is a something that is done throughout your project. If another workflow is used, such as squashing all your commits together before you merge your branch into master this is not necessary. Just make sure that your commit message for the squash commit includes a detailed message of the changes. A great tool for this case is a multiline commit message.

> For adding a new page in an example Project this could look like the following for separate commits:

Reviewer 5: ```suggestion
For adding a new page in an example project, this could look like the following for separate commits:
```

```
Add the Person model
Configure the Views to use the new Person Model
Use the Data from the Views in the index template
```

> and like this for a squashed multine commit:

Reviewer 5: ```suggestion
and like this for a squashed multiline commit:
```

```
Add Persons to the app

    - Add the Person model
    - Configure the Views to use the new Person Model
    - Use the Data from the Views in the index template
```

> For more information about how to write good commit messages you can check out [this great article by Chris Beams](https://chris.beams.io/posts/git-commit/)

Reviewer 5: ```suggestion
For more information about how to write good commit messages, you can check out [this great article by Chris Beams](https://chris.beams.io/posts/git-commit/)
```

## Rebasing

If possible, you should try to rebase changes that are requested during the Code Review phase into your previous commits.

> For more information on rebasing you can refer to the [official git documentation](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)

Reviewer 5: ```suggestion
For more information on rebasing, you can refer to the [official git documentation](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)
```
